src/mono/mono/mini/mini-llvm.c

```diff
@@ -10449,24 +10449,47 @@ MONO_RESTORE_WARNING
 			break;
 		}
 		case OP_WASM_SIMD_SWIZZLE: {
-			LLVMValueRef bidx = LLVMBuildBitCast (builder, rhs, LLVMVectorType (i1_t, 16), "");
+			LLVMValueRef bidx = rhs;
 			int nelems = LLVMGetVectorSize (LLVMTypeOf (lhs));
-			if (nelems < 16) {
-				int shift = nelems == 8 ? 1 : (nelems == 4 ? 2 : 3);
-				LLVMValueRef fill = LLVMConstNull (LLVMVectorType (i1_t, 16));
-				LLVMValueRef offset = LLVMConstNull (LLVMVectorType (i1_t, 16));
-				int stride = 16 / nelems;
-				for (int i = 0; i < nelems; ++i) {
-					for (int j = 0; j < stride; ++j) {
-						offset = LLVMBuildInsertElement (builder, offset, const_int8 (j), const_int8 (i * stride + j), "");
-						fill = LLVMBuildInsertElement (builder, fill, const_int8 (i * stride), const_int8 (i * stride + j), "");
+			int stride = 16 / nelems;
+
+			if (nelems != 16) {
+				bidx = LLVMBuildBitCast (builder, rhs, LLVMVectorType (i1_t, 16), "");
+				if (LLVMIsConstant(rhs)) {
+					LLVMValueRef indices [16];
+					for (int i = 0, k = 0; i < nelems; i++, k += stride) {
+						LLVMValueRef index = LLVMBuildExtractElement (builder, rhs, const_int32 (i), "");
+						g_assert (LLVMIsConstant (index));
+
+						uint64_t idx = LLVMConstIntGetZExtValue (index);
+
+						// handle invalid values by making sure they stay invalid
+						if (idx >= nelems)
+							idx = nelems;
```

Comment:
this seems to not do anything, is it meant to assign something other than nelems to idx?

Reply:
It assigns any invalid (too large) idx to a value that will still be invalid after the (idx * stride + j) and still fit within a byte so that the intrinsic will deal with it correctly.  It is a bit excessive but effectively free on this path.

Reply:
i64x2 with an index of 32 => 32 * 8 + 7 => 263.   (uint8)263 => 7